modules/tinymce/src/plugins/anchor/main/ts/core/Anchor.ts

```diff
@@ -5,33 +5,74 @@
  * For commercial licenses see https://www.tiny.cloud/
  */
 
+import { HTMLAnchorElement } from '@ephox/dom-globals';
 import Editor from 'tinymce/core/api/Editor';
-import { Element } from '@ephox/dom-globals';
+import RangeUtils from 'tinymce/core/api/dom/RangeUtils';
+import Tools from 'tinymce/core/api/util/Tools';
+import * as Settings from '../api/Settings';
+import * as Utils from './Utils';
 
-const isNamedAnchor = (editor: Editor, node: Element) => node.tagName === 'A' && editor.dom.getAttrib(node, 'href') === '';
+const removeEmptyNamedAnchorsInSelection = (editor: Editor) => {
+  const dom = editor.dom;
+  RangeUtils(dom).walk(editor.selection.getRng(), (nodes) => {
+    Tools.each(nodes, (node) => {
+      if (Utils.isEmptyNamedAnchor(node)) {
+        dom.remove(node, false);
+      }
+    });
+  });
+};
 
 const isValidId = (id: string) =>
   // Follows HTML4 rules: https://www.w3.org/TR/html401/types.html#type-id
   /^[A-Za-z][A-Za-z0-9\-:._]*$/.test(id);
 
+const getNamedAnchor = (editor: Editor): HTMLAnchorElement | null =>
+  editor.dom.getParent(editor.selection.getStart(), Utils.namedAnchorSelector) as HTMLAnchorElement;
+
 const getId = (editor: Editor) => {
-  const selectedNode = editor.selection.getNode();
-  return isNamedAnchor(editor, selectedNode) ? (selectedNode.getAttribute('id') || selectedNode.getAttribute('name')) : '';
+  const anchor = getNamedAnchor(editor);
+  if (anchor) {
+    return Utils.getIdFromAnchor(anchor);
+  } else {
+    return '';
+  }
 };
 
-const insert = (editor: Editor, id: string) => {
-  const selectedNode = editor.selection.getNode();
+const createAnchor = (editor: Editor, id: string): void => {
+  if (!Settings.isHtmlInNamedAnchorAllowed(editor)) {
+    editor.selection.collapse(true);
+  }
+  if (editor.selection.isCollapsed()) {
+    editor.insertContent(editor.dom.createHTML('a', { id }));
```

Comment:
The insertContent API executes the `mceInsertContent` command, which means it will get wrapped in the transaction. Would probably be clearer to move the transact to wrap all of this though and it shouldn't have any side effects.